### PR TITLE
feat: Add event trigger to handle job cleanup on table drop in vectorize schema

### LINKED ISSUE #148: drop table event triggers for vectorize.jobs
if a table is dropped that is part of vectorize.jobs, then the vectorize.job should also be deleted.

an event trigger can solve this. some scratch code as example of the idea

```sql
CREATE OR REPLACE FUNCTION after_drop_trigger()
RETURNS event_trigger AS $$
BEGIN
    DELETE FROM vectorize.job WHERE table_name = TG_TABLE_NAME;
END;
$$ LANGUAGE plpgsql;
```

```sql
CREATE EVENT TRIGGER trg_after_drop
ON sql_drop
WHEN TAG IN ('DROP TABLE')
EXECUTE FUNCTION after_drop_trigger();
```

### PR DESCRIPTION
**Pull request by akhilender-bongirwar**
fixes #148 
/claim #148

### CODE REVIEW

**Comment by akhilender-bongirwar** [comment]
@ChuckHend @jasonmp85, please review. I would be happy to incorporate any suggestions you have : )

**Comment by varshith257** [comment]
@akhilender-bongirwar another duplicate efforts with https://github.com/tembo-io/pg_vectorize/pull/164

PS: It's getting reviewed though 

I got occupied with other works recently and didn't made time to push to endline and complete PR. Anyhow will complete it now.

**Comment by akhilender-bongirwar** [comment]
> @akhilender-bongirwar another duplicate efforts with #164
> 
> PS: It's getting reviewed though
> 
> I got occupied with other works recently and didn't made time to push to endline and complete PR. Anyhow will complete it now.

@varshith257  I saw your PR was inactive for 5 days and I had an idea for resolving the issue.

**Comment by akhilender-bongirwar** [comment]
@ChuckHend, could you also review my implementation?

I have a question: are all the integration tests ignored by default? If so, how are we writing and running tests to verify the changes?

**Comment by ChuckHend** [comment]
All of the CI was paused since this PR is from a branch outside the `tembo-io` org. The unit and integration tests from this PR are running here https://github.com/tembo-io/pg_vectorize/actions/runs/11631847410

You will also need to add a test for this feature. Keep in mind https://github.com/tembo-io/pg_vectorize/pull/164/files was opened first, so we are reviewing that first.

**Comment by akhilender-bongirwar** [comment]
> All of the CI was paused since this PR is from a branch outside the `tembo-io` org. The unit and integration tests from this PR are running here https://github.com/tembo-io/pg_vectorize/actions/runs/11631847410
> 
> You will also need to add a test for this feature. Keep in mind https://github.com/tembo-io/pg_vectorize/pull/164/files was opened first, so we are reviewing that first.

Sure, I understand : )

File: extension/sql/meta.sql [hunk 1]
```diff
@@ -20,6 +20,24 @@ GRANT SELECT ON ALL SEQUENCES IN SCHEMA vectorize TO pg_monitor;
...
 
+CREATE OR REPLACE FUNCTION handle_table_drop() 
+RETURNS event_trigger AS $$
+DECLARE
+    obj RECORD;
+BEGIN
+    FOR obj IN SELECT * FROM pg_event_trigger_dropped_objects() LOOP
+        IF obj.object_type = 'table' AND obj.object_schema = 'vectorize' THEN
+            DELETE FROM vectorize.job 
+            WHERE name = obj.object_name;
```
Comment (ChuckHend) [inline extension/sql/meta.sql:31]: The tables that are dropped could be in any schema. And the record on the vectorize.job table that needs to be dropped points to that table in the params ->> 'table' and params ->> 'schema' values.

Comment (akhilender-bongirwar) [reply extension/sql/meta.sql]: yeah 😅

**Comment by akhilender-bongirwar** [comment]
image

@ChuckHend, please approve workflows once.

File: extension/sql/meta.sql [hunk 2]
```diff
@@ -20,6 +20,39 @@ GRANT SELECT ON ALL SEQUENCES IN SCHEMA vectorize TO pg_monitor;
...
+            RAISE NOTICE 'Event Trigger Fired for table drop: %, schema: %', table_name, schema_name;
+            
+            -- Perform cleanup: delete the associated job from the vectorize.job table
+            DELETE FROM vectorize.job
+            WHERE params ->> 'table' = table_name
+            AND params ->> 'schema' = schema_name;
+            
+            IF NOT FOUND THEN
+                RAISE NOTICE 'No matching job found for table: %, schema: %', table_name, schema_name;
+            END IF;
```
Comment (ChuckHend) [inline extension/sql/meta.sql:44]: Will this notice fire on all `DROP TABLE` in the database?

Comment (akhilender-bongirwar) [reply extension/sql/meta.sql:44]: I've used the RAISE NOTICE statements for debugging purposes to confirm that the event trigger is firing correctly. However, in production, these notices are not necessary, so I'll remove them.

Comment (akhilender-bongirwar) [reply extension/sql/meta.sql]: Done.

**Comment by akhilender-bongirwar** [comment]
@ChuckHend, There are some failing [tests](https://github.com/tembo-io/pg_vectorize/actions/runs/12240904372/job/34145932283), but I didn’t alter them, and they are not related to the changes I made. Please let me know how you would like to proceed.

**Comment by ChuckHend** [comment]
Do not worry about `test_cohere` and `test_private_hf_model`, since those integration tests will not pass when they are run outside of the tembo org.

I'll have this reviewed within the next couple days! Thank you.

**Comment by akhilender-bongirwar** [comment]
> Do not worry about `test_cohere` and `test_private_hf_model`, since those integration tests will not pass when they are run outside of the tembo org.
> 
> I'll have this reviewed within the next couple days! Thank you.

ok. Thank you ❤️

**Review by ChuckHend — APPROVED** [review]

**Comment by ChuckHend** [comment]
Looks great! @varshith257 had a good start on this bounty early and there was clearly value added most recently in this PR from @akhilender-bongirwar. From what I've observed, there were some complementary efforts and I think it would be fair to split the bounty. I will split the bounty as soon as I can figure out how to do that.

**Pull request merged**